clusterctl/clusterdeployer/clusterdeployer.go

```diff
@@ -1,38 +1,392 @@
+	if err != nil {
+		return nil, fmt.Errorf("unable to create external client: %v", err)
+	}
+	return externalClient, nil
+}
+
+func (d *ClusterDeployer) createInternalCluster(externalClient ClusterClient) (ClusterClient, error) {
+	cluster, master, _, err := getClusterAPIObjects(externalClient)
+	if err != nil {
+		return nil, err
+	}
+	glog.V(1).Info("Getting internal cluster kubeconfig.")
+	internalKubeconfig, err := waitForKubeconfigReady(d.provider, cluster, master)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get internal cluster kubeconfig: %v", err)
+	}
+	err = d.writeKubeconfig(internalKubeconfig)
+	if err != nil {
+		return nil, err
+	}
+	defer glog.Info("You can now access your cluster with kubectl --kubeconfig %v", d.kubeconfigOutput)
+	internalClient, err := d.clientFactory.ClusterClient(internalKubeconfig)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create internal cluster client: %v", err)
+	}
+	return internalClient, nil
+}
+
+func (d *ClusterDeployer) updateClusterEndpoint(client ClusterClient) error {
+	// Update cluster endpoint. Needed till this logic moves into cluster controller.
+	// TODO: https://github.com/kubernetes-sigs/cluster-api/issues/158
+	// Fetch fresh objects.
+	cluster, master, _, err := getClusterAPIObjects(client)
+	if err != nil {
+		return err
+	}
+	masterIP, err := d.provider.GetIP(cluster, master)
+	if err != nil {
+		return fmt.Errorf("unable to get master IP: %v", err)
+	}
+	err = client.UpdateClusterObjectEndpoint(masterIP)
+	if err != nil {
+		return fmt.Errorf("unable to update cluster endpoint: %v", err)
+	}
+	return nil
+}
+
+func (d *ClusterDeployer) applyClusterAPIStack(client ClusterClient) error {
+	glog.V(2).Info("Applying Cluster API APIServer.")
+	err := d.applyClusterAPIApiserver(client)
+	if err != nil {
+		return fmt.Errorf("unable to apply cluster apiserver: %v", err)
+	}
+	glog.V(2).Info("Applying Cluster API Provider Components.")
+	err = d.applyClusterAPIControllers(client)
+	if err != nil {
+		return fmt.Errorf("unable to apply cluster api controllers: %v", err)
+	}
+	return nil
+}
+
+func (d *ClusterDeployer) applyClusterAPIStackWithPivoting(client ClusterClient, source ClusterClient) error {
+	glog.V(1).Info("Applying Cluster API stack.")
+	glog.V(2).Info("Applying Cluster API API server.")
+	err := d.applyClusterAPIApiserver(client)
+	if err != nil {
+		return fmt.Errorf("unable to apply cluster api apiserver: %v", err)
+	}
+
+	glog.V(2).Info("Pivoting Cluster API objects.")
+	err = pivot(source, client)
+	if err != nil {
+		return fmt.Errorf("unable to pivot cluster API objects: %v", err)
+	}
+
+	glog.V(2).Info("Applying Cluster API Provider Components.")
+	err = d.applyClusterAPIControllers(client)
+	if err != nil {
+		return fmt.Errorf("unable to apply cluster api controllers: %v", err)
+	}
+
+	return nil
+}
+
+func (d *ClusterDeployer) applyClusterAPIApiserver(client ClusterClient) error {
+	yaml, err := getApiServerYaml()
```

Comment:
Nit: I would prefer if this was called getClusterApiServerYaml since it could mean the core api server.

Reply:
I think whether it uses yaml under the covers or something else is a implementation detail. I was hoping the clusterapi + apiserver in the method name would make it clear it is the cluster api apiserver as opposed to the main apiserver. Is this ok or do you have other ideas on how do distinguish between the main apiserver and the clusterapi apiserver for the method name?

Reply:
Ok, that makes sense!